sdk/storage/azure-storage-blob/azure/storage/blob/_shared/authentication.py

```diff
@@ -48,6 +48,33 @@ def _wrap_exception(ex, desired_type):
     # TODO: In the future we will log the trace
     return desired_type('{}: {}'.format(ex.__class__.__name__, msg))
 
+# This method attempts to emulate the sorting done by the service
+@staticmethod
+def _service_sort(input_header):
+    # Define the custom alphabet for weights
+    custom_weights = "-!#$%&*.^_|~+\"\'(),/`~0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]abcdefghijklmnopqrstuvwxyz{}"
+
+    # Build dict of tuples and list of keys
+    header_dict = dict()
+    header_keys = []
+
+    for k, v in input_header:
+        header_dict[k] = v
+        header_keys.append(k)
+
+    # Sort according to custom defined weights
+    try:
+        header_keys = sorted(header_keys, key=lambda word: [custom_weights.index(c) for c in word])
+    except ValueError:
+        raise ValueError("Illegal character encountered when sorting headers.")
```

Comment:
Open to suggestions for what error to raise. Although it already was raising a `ValueError` even without putting this logic here, the error was rather useless to users as it could be considered sort of cryptic, but what it really meant was that a character it attempted to sort is not in our custom alphabet.

Sample error without this logic:
```
>   header_keys = sorted(header_keys, key=lambda word: [custom_weights.index(c) for c in word])
E   ValueError: substring not found
```